service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceTwin.java

```diff
@@ -259,6 +260,72 @@ public void replaceTags(DeviceTwinDevice device) throws UnsupportedOperationExce
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Replace the full device twin for a given device with the provided device twin.
+     * @param device The device twin object to replace the current device twin object.
+     * @throws IotHubException If any an IoT Hub level exception is thrown. For instance,
+     * if the request is unauthorized, a exception that extends IotHubException will be thrown.
+     * @throws IOException If the request failed to send to IoT Hub.
+     * @return The Twin object's current state returned from the service after the replace operation.
+     */
+    public DeviceTwinDevice replaceTwin(DeviceTwinDevice device) throws IotHubException, IOException
+    {
+        if (device == null || device.getDeviceId() == null || device.getDeviceId().length() == 0)
+        {
+            throw new IllegalArgumentException("Instantiate a device and set device id to be used");
+        }
+
+        URL url;
+        if ((device.getModuleId() == null) || device.getModuleId().length() ==0)
+        {
+            url = this.iotHubConnectionString.getUrlTwin(device.getDeviceId());
+        }
+        else
+        {
+            url = this.iotHubConnectionString.getUrlModuleTwin(device.getDeviceId(), device.getModuleId());
+        }
+
+        TwinState twinState = new TwinState(device.getTagsMap(), device.getDesiredMap(), null);
+        String twinJson = twinState.toJsonElement().toString();
+
+        Proxy proxy = options.getProxyOptions() != null ? options.getProxyOptions().getProxy() : null;
+
+        HttpResponse httpResponse = DeviceOperations.request(
+            this.iotHubConnectionString,
+            url,
+            HttpMethod.PUT,
+            twinJson.getBytes(StandardCharsets.UTF_8),
+            String.valueOf(requestId++),
```

Comment:
Interesting approach for request Id

Reply:
Yeah, this is copied from the other methods. I don't care for it that much, to be honest.

Reply:
the tone was lost in my comment above - :neutral_face:

Reply:
I may revisit how we handle request Ids at a later time